File: src/main/scala/nl/rug/jbi/search/ssap/util/ProjectContainer.scala
package nl.rug.jbi.search.ssap.util

import java.io.{InputStream, File}

/**
  * Represents the container of a Java project, from which one can access the project's .class files.
  *
  * @author Daniel Feitosa
  */
abstract class ProjectContainer(val project: File) extends AutoCloseable {
  if(!isValid)
    throw new IllegalArgumentException(s"'$project' isn't a valid project")

  def isValid: Boolean
  def forEachClass[A](callback: (InputStream) => A)
  def getClassStream(classname: String): InputStream
}

File: ssap/src/main/java/nl/rug/jbi/search/ssap/util/ProjectContainer.java
package nl.rug.jbi.search.ssap.util;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;


public abstract class ProjectContainer implements AutoCloseable {

    public final File project;

    public ProjectContainer(File project) {
        this.project = project;

        if(!isValid()) {
            throw new IllegalArgumentException(project + " isn't a valid project");
        }
    }

    public abstract Boolean isValid();

    public abstract <A> void forEachClass (CallbackFunction<A> callback) throws IOException;

    public abstract InputStream getClassStream(String classname) throws IOException;
}
